`&&self` suggestion from compiler

### Code

struct SomeStruct;

impl SomeStruct {
    fn some_fn(&some_name) {
        self
    }
}

fn main() {}

### Current output

error[E0424]: expected value, found module `self`
 --> src/main.rs:5:9
  |
4 |     fn some_fn(&some_name) {
  |        ------- this function doesn't have a `self` parameter
5 |         self
  |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
  |
help: add a `self` receiver parameter to make the associated `fn` a method
  |
4 |     fn some_fn(&&self, some_name) {
  |                 ++++++

For more information about this error, try `rustc --explain E0424`.

### Desired output

error[E0424]: expected value, found module `self`
 --> src/main.rs:5:9
  |
4 |     fn some_fn(&some_name) {
  |        ------- this function doesn't have a `self` parameter
5 |         self
  |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
  |
help: add a `self` receiver parameter to make the associated `fn` a method
  |
4 |     fn some_fn(self, &some_name) {
  |                +++++

For more information about this error, try `rustc --explain E0424`.

### Rationale and extra context

Maybe the user forgot to add the `self` parameter in the code. Or maybe they forgot to rename `some_name` to `self` while refactoring the code. Either way, for beginners, the compiler suggesting to use `&&self` instead of `self` or `&self` may be confusing.

### Other cases

_No response_

### Rust Version

rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: aarch64-apple-darwin
release: 1.81.0
LLVM version: 18.1.7

### Anything else?

_No response_

Taking the example code and building with:
```
cargo 1.96.0 (30a34c682 2026-05-25)
release: 1.96.0
commit-hash: 30a34c6821b57de0aaec83a901aca39f88f6778c
commit-date: 2026-05-25
host: aarch64-apple-darwin
libgit2: 1.9.2 (sys:0.20.4 vendored)
libcurl: 8.7.1 (sys:0.4.87+curl-8.19.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 3.5.4 30 Sep 2025
os: Mac OS 26.5.1 [64-bit]
```
The resultant output is:
```
   Compiling aaself v0.1.0 (/Users/rod/code/rust/triage/2026/06/aaself)
error: expected one of `:`, `@`, or `|`, found `)`
 --> src/main.rs:4:26
  |
4 |     fn some_fn(&some_name) {
  |                          ^ expected one of `:`, `@`, or `|`
  |
help: if this is a `self` type, give it a parameter name
  |
4 |     fn some_fn(self: &some_name) {
  |                +++++
help: if this is a parameter name, give it a type
  |
4 -     fn some_fn(&some_name) {
4 +     fn some_fn(some_name: &TypeName) {
  |
help: if this is a type, explicitly ignore the parameter name
  |
4 |     fn some_fn(_: &some_name) {
  |                ++

error[E0424]: expected value, found module `self`
 --> src/main.rs:5:9
  |
4 |     fn some_fn(&some_name) {
  |        ------- this function doesn't have a `self` parameter
5 |         self
  |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
  |
help: add a `self` receiver parameter to make the associated `fn` a method
  |
4 |     fn some_fn(&&self, some_name) {
  |                 ++++++

For more information about this error, try `rustc --explain E0424`.
error: could not compile `aaself` (bin "aaself") due to 2 previous errors
```
The suggestion &&self is still made, but now there are two other suggestions. So progress but &&self is still not a very good suggestion.

--- a/compiler/rustc_parse/src/parser/item.rs
+++ b/compiler/rustc_parse/src/parser/item.rs
@@ -3410,6 +3410,7 @@ impl<'a> Parser<'a> {
                 let (pat, colon) = this.parse_fn_param_pat_colon()?;
                 if !colon {
                     let mut err = this.unexpected().unwrap_err();
+                    let pat_span = pat.span;
                     return if let Some(ident) = this.parameter_without_type(
                         &mut err,
                         pat,
@@ -3418,7 +3419,9 @@ impl<'a> Parser<'a> {
                         fn_parse_mode,
                     ) {
                         let guar = err.emit();
-                        Ok((dummy_arg(ident, guar), Trailing::No, UsePreAttrPos::No))
+                        let mut arg = dummy_arg(ident, guar);
+                        arg.span = pat_span;
+                        Ok((arg, Trailing::No, UsePreAttrPos::No))
                     } else {
                         Err(err)
                     };

--- a/tests/ui/suggestions/suggest-add-self-issue-131084.rs
+++ b/tests/ui/suggestions/suggest-add-self-issue-131084.rs
@@ -11,6 +11,18 @@ impl SomeStruct {
         self
         //~^ ERROR expected value, found module `self`
     }
+
+    fn mut_param(mut some_name) {
+        //~^ ERROR expected one of `:`, `@`, or `|`, found `)`
+        self
+        //~^ ERROR expected value, found module `self`
+    }
+
+    fn type_before_name(String s) {
+        //~^ ERROR expected one of `:`, `@`, or `|`, found `s`
+        self
+        //~^ ERROR expected value, found module `self`
+    }
 }
 
 fn main() {}

--- a/tests/ui/suggestions/suggest-add-self-issue-131084.rs
+++ b/tests/ui/suggestions/suggest-add-self-issue-131084.rs
@@ -1,0 +1,16 @@
+//@ check-fail
+
+// A recovered `&param` should not make the missing-`self` suggestion insert
+// the receiver after the leading `&`.
+
+struct SomeStruct;
+
+impl SomeStruct {
+    fn some_fn(&some_name) {
+        //~^ ERROR expected one of `:`, `@`, or `|`, found `)`
+        self
+        //~^ ERROR expected value, found module `self`
+    }
+}
+
+fn main() {}

--- a/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
+++ b/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
@@ -18,6 +18,34 @@ help: if this is a type, explicitly ignore the parameter name
 LL |     fn some_fn(_: &some_name) {
    |                ++
 
+error: expected one of `:`, `@`, or `|`, found `)`
+  --> $DIR/suggest-add-self-issue-131084.rs:15:31
+   |
+LL |     fn mut_param(mut some_name) {
+   |                               ^ expected one of `:`, `@`, or `|`
+   |
+help: if this is a `self` type, give it a parameter name
+   |
+LL |     fn mut_param(self: mut some_name) {
+   |                  +++++
+help: if this is a parameter name, give it a type
+   |
+LL |     fn mut_param(mut some_name: TypeName) {
+   |                               ++++++++++
+help: if this is a type, explicitly ignore the parameter name
+   |
+LL |     fn mut_param(_: mut some_name) {
+   |                  ++
+
+error: expected one of `:`, `@`, or `|`, found `s`
+  --> $DIR/suggest-add-self-issue-131084.rs:21:32
+   |
+LL |     fn type_before_name(String s) {
+   |                         -------^
+   |                         |      |
+   |                         |      expected one of `:`, `@`, or `|`
+   |                         help: declare the type after the parameter binding: `<identifier>: <type>`
+
 error[E0424]: expected value, found module `self`
   --> $DIR/suggest-add-self-issue-131084.rs:11:9
    |
@@ -32,6 +60,34 @@ help: add a `self` receiver parameter to make the associated `fn` a method
 LL |     fn some_fn(&self, &some_name) {
    |                ++++++
 
-error: aborting due to 2 previous errors
+error[E0424]: expected value, found module `self`
+  --> $DIR/suggest-add-self-issue-131084.rs:17:9
+   |
+LL |     fn mut_param(mut some_name) {
+   |        --------- this function doesn't have a `self` parameter
+LL |
+LL |         self
+   |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
+   |
+help: add a `self` receiver parameter to make the associated `fn` a method
+   |
+LL |     fn mut_param(&self, mut some_name) {
+   |                  ++++++
+
+error[E0424]: expected value, found module `self`
+  --> $DIR/suggest-add-self-issue-131084.rs:23:9
+   |
+LL |     fn type_before_name(String s) {
+   |        ---------------- this function doesn't have a `self` parameter
+LL |
+LL |         self
+   |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
+   |
+help: add a `self` receiver parameter to make the associated `fn` a method
+   |
+LL |     fn type_before_name(&self, String s) {
+   |                         ++++++
+
+error: aborting due to 6 previous errors
 
 For more information about this error, try `rustc --explain E0424`.

--- a/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
+++ b/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
@@ -29,8 +29,8 @@ LL |         self
    |
 help: add a `self` receiver parameter to make the associated `fn` a method
    |
-LL |     fn some_fn(&&self, some_name) {
-   |                 ++++++
+LL |     fn some_fn(&self, &some_name) {
+   |                ++++++
 
 error: aborting due to 2 previous errors
 

--- a/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
+++ b/tests/ui/suggestions/suggest-add-self-issue-131084.stderr
@@ -1,0 +1,37 @@
+error: expected one of `:`, `@`, or `|`, found `)`
+  --> $DIR/suggest-add-self-issue-131084.rs:9:26
+   |
+LL |     fn some_fn(&some_name) {
+   |                          ^ expected one of `:`, `@`, or `|`
+   |
+help: if this is a `self` type, give it a parameter name
+   |
+LL |     fn some_fn(self: &some_name) {
+   |                +++++
+help: if this is a parameter name, give it a type
+   |
+LL -     fn some_fn(&some_name) {
+LL +     fn some_fn(some_name: &TypeName) {
+   |
+help: if this is a type, explicitly ignore the parameter name
+   |
+LL |     fn some_fn(_: &some_name) {
+   |                ++
+
+error[E0424]: expected value, found module `self`
+  --> $DIR/suggest-add-self-issue-131084.rs:11:9
+   |
+LL |     fn some_fn(&some_name) {
+   |        ------- this function doesn't have a `self` parameter
+LL |
+LL |         self
+   |         ^^^^ `self` value is a keyword only available in methods with a `self` parameter
+   |
+help: add a `self` receiver parameter to make the associated `fn` a method
+   |
+LL |     fn some_fn(&&self, some_name) {
+   |                 ++++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0424`.

